checks.d/dns_check.py

```diff
@@ -0,0 +1,73 @@
+# stdlib
+import time
+
+# 3p
+import dns.resolver
+
+# project
+from checks import AgentCheck
+
+
+class DNSCheck(AgentCheck):
+    SERVICE_CHECK_NAME = 'dns.can_resolve'
+    DEFAULT_TIMEOUT = 5
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        AgentCheck.__init__(self, name, init_config, agentConfig, instances)
+        self.default_timeout = init_config.get('default_timeout', self.DEFAULT_TIMEOUT)
+
+    def check(self, instance):
+        if 'hostname' not in instance:
+            self.log.info("Skipping instance, no hostname found.")
+            return
+
+        timeout = float(instance.get('timeout', self.default_timeout))
+        tags = self._get_tags(instance)
+
+        timing = self.query_dns(instance, timeout)
+        if timing > 0:
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=tags)
+            self.gauge('dns.response_time', timing, tags=tags)
+
+    def query_dns(self, instance, timeout):
+        """Perform the DNS query, time it and return the measure"""
+        start_time = end_time = 0.0
+        resolver = dns.resolver.Resolver()
+        resolver.lifetime = timeout
+
+        hostname = instance.get('hostname')
+
+        # if a specific DNS server was defined use it, else use the system default
+        nameserver = instance.get('nameserver')
+        if nameserver is not None:
+            resolver.nameservers = [nameserver]
+
+        start_time = time.time()
+        try:
+            answer = resolver.query(hostname)
+            assert(answer.rrset.items[0].address)
+            end_time = time.time()
+        except dns.exception.Timeout:
+            self.log.error('DNS resolution of %s timed out' % hostname)
+            self.service_check(self.SERVICE_CHECK_NAME,
+                               AgentCheck.CRITICAL, tags=self._get_tags(instance))
+            return -1
+        except Exception as ex:
+            self.log.error('DNS resolution of %s failed.' % hostname)
+            self.log.exception(ex)
+            self.service_check(self.SERVICE_CHECK_NAME,
+                               AgentCheck.CRITICAL, tags=self._get_tags(instance))
+            return -1
```

Comment:
One additional thought: `AgentCheck`s usually raise on top of sending a `CRITICAL` service check.